pkg/wireguard/operator/operator.go

```diff
@@ -0,0 +1,369 @@
+	default:
+		return errors.New("unsupported family")
+	}
+
+	nodeName := n.ObjectMeta.Name
+	allocated := false
+	defer func() {
+		if allocated {
+			log.WithFields(logrus.Fields{
+				logfields.NodeName: nodeName,
+				logfields.IPAddr:   ipByNode[nodeName],
+			}).Info("Allocated wireguard IP")
+		}
+	}()
+
+	ip, found := findWireguardIP(n, family)
+
+	if o.restoring && !found {
+		// We will allocate an IP once we have learned about all previously
+		// allocated IPs (after sync with k8s has finished).
+		ipAllocAfterRestore[nodeName] = struct{}{}
+		return nil
+	}
+
+	if !found {
+		// No IP was found in CiliumNode, so let's allocate one
+
+		var ip net.IP
+		var err error
+		if prevIP, ok := ipByNode[nodeName]; ok {
+			ip = prevIP
+			// Previously, the node had an IP assigned to it, so let's reallocate
+			// it. This can happen when someone manually removes the wireguard IP
+			// from CiliumNode object.
+			err = ipAlloc.Allocate(prevIP)
+			if err != nil && !errors.Is(err, ipallocator.ErrAllocated) {
+				return fmt.Errorf("failed to re-allocate IP addr for node %s: %w", nodeName, err)
+			}
+		} else {
+			ip, err = ipAlloc.AllocateNext()
+			if err != nil {
+				return fmt.Errorf("failed to allocate IP addr for node %s: %w", nodeName, err)
+			}
+		}
+
+		if err := o.setCiliumNodeIP(nodeName, ip); err != nil {
+			ipAlloc.Release(ip)
+			return err
+		}
+		ipByNode[nodeName] = ip
+		allocated = true
+
+		return nil
+	}
+
+	// An IP was found in CiliumNode. This could happen in both states
+	// (restoring and after restoring).
+	if prevIP, ok := ipByNode[nodeName]; ok {
+		if !prevIP.Equal(ip) {
+			// The IP we previously learnt does not match. Release it first before
+			// we reallocate the IP from CiliumNode
+			ipAlloc.Release(prevIP)
+			delete(ipByNode, nodeName)
+
+			if err := ipAlloc.Allocate(ip); err != nil {
+				return fmt.Errorf("failed to re-allocate IP addr %s for node %s: %w", ip, nodeName, err)
+			}
+			ipByNode[nodeName] = ip
+		}
+	} else {
+		// We don't know about this IP, let's allocate it (can happen during
+		// restore).
+		if err := ipAlloc.Allocate(ip); err != nil {
+			return err
+		}
+		ipByNode[nodeName] = ip
+	}
+
+	return nil
+}
+
+func (o *Operator) setCiliumNodeIP(nodeName string, ip net.IP) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		node, err := o.ciliumNodeUpdater.Get(nodeName)
+		if err != nil {
+			return err
+		}
+
+		for _, addr := range node.Spec.Addresses {
+			if addr.Type == addressing.NodeWireguardIP {
+				if foundIP := net.ParseIP(addr.IP); foundIP != nil && foundIP.Equal(ip) {
+					return nil
+				}
+			}
+		}
+
+		node.Spec.Addresses = append(node.Spec.Addresses, v2.NodeAddress{Type: addressing.NodeWireguardIP, IP: ip.String()})
+		_, err = o.ciliumNodeUpdater.Update(nil, node)
+		return err
+	})
+
+	log.WithFields(logrus.Fields{
+		logfields.NodeName: nodeName,
+		logfields.IPAddr:   ip,
+	}).Info("Set wireguard IP")
```

Comment:
how about adding the error here as well?

Or maybe something like this? (totally untested)
```
	l := log.WithFields(logrus.Fields{
		logfields.NodeName: nodeName,
		logfields.IPAddr:   ip,
	})

	if err != nil {
		l.Info("Set wireguard IP")
	} else {
		l.WithError(err).Warn("Set wireguard IP failed")
	}

	return err
```